janus_server/src/datastore.rs

```diff
@@ -0,0 +1,318 @@
+    #[error("decoding error: {0}")]
+    DecodeError(#[from] CodecError),
+    /// An arbitrary error returned from the user callback; unrelated to DB internals. This error
+    /// will never be generated by the datastore library itself.
+    #[error(transparent)]
+    User(#[from] Box<dyn std::error::Error + Send + Sync>),
+}
+
+impl Error {
+    // is_serialization_failure determines if a given error corresponds to a Postgres
+    // "serialization" failure, which requires the entire transaction to be aborted & retried from
+    // the beginning per https://www.postgresql.org/docs/9.5/transaction-iso.html.
+    fn is_serialization_failure(&self) -> bool {
+        match self {
+            // T_R_SERIALIZATION_FAILURE (40001) is documented as the error code which is always used
+            // for serialization failures which require rollback-and-retry.
+            Error::Db(err) => err
+                .code()
+                .map_or(false, |c| c == &SqlState::T_R_SERIALIZATION_FAILURE),
+            _ => false,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    // TODO(brandon): use podman instead of docker for container management once testcontainers supports this
+
+    use crate::message::{ExtensionType, HpkeConfigId};
+
+    use super::*;
+    use lazy_static::lazy_static;
+    use std::str;
+    use testcontainers::{clients, images::postgres::Postgres, Container, Docker};
+    use tokio_postgres::NoTls;
+
+    lazy_static! {
+        static ref SCHEMA: &'static str = {
+            let schema_bytes = include_bytes!("../../db/schema.sql");
+            str::from_utf8(schema_bytes).unwrap()
+        };
+    }
+
+    /// ephemeral_db creates a new Datastore instance backed by an ephemeral database which has the
+    /// Janus schema applied but is otherwise empty.
+    ///
+    /// Dropping the second return value causes the database to be shut down & cleaned up.
+    pub(crate) async fn ephemeral_db<D: Docker>(
+        container_client: &D,
+    ) -> (Datastore, Container<'_, D, Postgres>) {
+        // Start an instance of Postgres running in a container.
+        let db_container = container_client.run(Postgres::default().with_version(14));
+
+        // Connect to our new instance of Postgres, retrieve a client, and start servicing the
+        // connection.
+        const POSTGRES_DEFAULT_PORT: u16 = 5432;
+        let connection_string = format!(
+            "postgres://postgres:postgres@localhost:{}/postgres",
+            db_container.get_host_port(POSTGRES_DEFAULT_PORT).unwrap()
+        );
+        let (postgres_client, conn) = tokio_postgres::connect(&connection_string, NoTls)
+            .await
+            .unwrap();
+        tokio::spawn(async move { conn.await.unwrap() });
+
+        // Run our schema on the new database.
+        postgres_client.batch_execute(*SCHEMA).await.unwrap();
+        (Datastore::new(postgres_client), db_container)
+    }
+
+    #[tokio::test]
+    async fn roundtrip_report() {
```

Comment:
I suggest installing a `tracing` subscriber in tests or you won't ever see `info!`, `err!` etc. output. The obnoxious thing is that you can only install a single subscriber per process, so you have to guard it with an `std::sync::Once`, as I did in the `src/aggregator.rs` tests. This might be a good time to place the test subscriber installation logic into a shared `tests` module and use that here. However we could also leave that for a future PR; a TODO here will suffice.